✨ Support dynamic version and description for Python parser

dd opened the pull request:
Updated the Python parser:

- now supports `dynamic` fields (`version` and `description`);
- if `version` is dynamic — it tries to resolve it using `hatch`, `flit`, `pdm`, or `setuptools_scm`;
- if `description` is dynamic — it reads it from the `README`;

I added [smol-toml](https://www.npmjs.com/package/smol-toml) because it supports the TOML 1.0.0 spec, unlike the [toml](https://www.npmjs.com/package/toml) package which only supports up to 0.4.0. That said, I kept `toml` in place for other parsers — though maybe it’s worth replacing it everywhere eventually.

Also, for dynamic `version` fields in `pyproject.toml`, I opted to run an external command to retrieve the version, depending on the build backend. For example: `hatch version` for Hatch, `flit info` for Flit, etc. I figured it’s simpler and more robust than trying to parse the version manually from the source files.

Also added tests for all of that.

The parser is now more in line with PEP 621 and modern `pyproject.toml` setups.

dd committed:
✨ Add support for dynamic fields for python parser

dd committed:
🧪 Add tests for dynamic fields and refine pyproject parser

dd committed:
🚨 Linting code

dd commented:
Hmm, smol-toml, is only available starting from Node.js 18...
How much of a problem would that be? Node 12 does seem quite outdated at this point.

frinyvonnick commented:
Hi @dd 👋 Thank you for your contribution 🙇 

> Hmm, smol-toml, is only available starting from Node.js 18... How much of a problem would that be? Node 12 does seem quite outdated at this point.

Maybe you can start by upgrading to Node.js LTS version?

dd commented:
Alright, I’ll try to update Node.js, hopefully sometime around the weekend!

dd committed:
👷 Upgrade Node to version 18-alpine

dd committed:
🔧 Update engines.node to require Node.js >=18

dd committed:
🔧 Update Node version in GH Actions workflow

dd committed:
🔧 Update Node version to LTS

dd committed:
🔧 Specify Node LTS version using .nvmrc

dd commented:
@frinyvonnick I updated the Node.js version to the latest LTS wherever applicable and added a .nvmrc file for local development and Netlify compatibility. Also set the minimum required Node version to 18 in package.json.

Hope I didn’t miss anything! 😊

frinyvonnick commented:
Thank you @dd 🙇  I'll do the review as soon as possible 👍

frinyvonnick commented:
@dd locally everything is fine but the CI does not pass for some reason. Can you give it a look? I enable it to run on this pull request 😄

dd committed:
🩹 Fixed Node versions in CI/CD configurations

dd commented:
@frinyvonnick Yes, it was my mistake — I've fixed it! 😅

Btw, I also updated the tests to include all the latest LTS Node versions. Before, it was using just one version. Not entirely sure how much it mattered though...

frinyvonnick commented:
@dd can you update the documentation here https://github.com/frinyvonnick/gitmoji-changelog/blob/master/packages/gitmoji-changelog-documentation/README.md to add a paragraph of explanation about what you did in the python section, please?

frinyvonnick commented on packages/gitmoji-changelog-cli/src/presets/python.js:
```diff
@@ -1,29 +1,60 @@
-const toml = require('toml')
 const fs = require('fs')
+const toml = require('smol-toml')
```
So **toml** dependency is not used anymore, we could remove it, right?

dd replied:
**toml** package is still used in the Cargo preset, but it can be replaced with **smol-toml** there as well (tests are passing).

frinyvonnick replied:
Oh I missed it, thx 🙏

frinyvonnick commented:
Aside from my comments about the documentation and the unused dependency it looks good to me 👍

dd commented:
@frinyvonnick 

Hi, sorry for the long reply.

I’ve decided to close this PR because the initial approach (adding support for specific Python project managers) turned out to be too narrow.
In the [new PR](https://github.com/frinyvonnick/gitmoji-changelog/pull/280), there’s no dependency on any particular package manager - instead, you can use the `--version-command` option to specify a command that returns the project version. This is more flexible and, I believe, the better approach.

I’ve also updated the documentation and included all relevant changes from this PR. Please check out the new PR when you have a chance.

If possible, I’d also appreciate a review of my other PRs:

* [Bugfix for CLI options](https://github.com/frinyvonnick/gitmoji-changelog/pull/278)
* [Option to ignore merges in changelog generation](https://github.com/frinyvonnick/gitmoji-changelog/pull/277)
* [Option to consider only stable versions when generating the changelog](https://github.com/frinyvonnick/gitmoji-changelog/pull/279)
* [Small visual update for the changelog template](https://github.com/frinyvonnick/gitmoji-changelog/pull/276)

Thanks!
